### ISSUE TITLE
How to modify mqtt example without secure connection -- Question

### ISSUE THREAD

**Issue by RudyFiero** [lubosmato/esp32-essentials#7]
Hi Luboš

In the other issue I said I was having connecting problems. I am sure the problem I was having was related to the router I was using. I had tested it with different code and also had unreliable connections. The odd thing was that I would get a event confirming the connection to the router, but the WL_CONNECTED would not change state, and it would be stuck in the while loop. 

My question now is, what do I need to do to use the mqtt example without encryption? 

I had looked at Espressif's mqtt example and compared it with their mqtt-ssl example. There was little difference, so I tried to do changes to your code. Unfortunately I was unsuccessful. I get stack overflow errors when connection of WiFi is called. 

I had commented out the following lines.

```
// don't forget to link root certificate in root CMakeLists.txt:
// target_add_binary_data(${CMAKE_PROJECT_NAME}.elf "main/cert.pem" TEXT)
// see how to generate certificate below in Details section of readme.md
////extern const uint8_t mqttCertBegin[] asm("_binary_cert_pem_start");
////extern const uint8_t mqttCertEnd[] asm("_binary_cert_pem_end");


 ///auto mqttCert =
 /// std::string_view{reinterpret_cast<const char*>(mqttCertBegin), std::size_t(mqttCertEnd - mqttCertBegin)};
```

I removed mqttCert in the following.

`  es::Mqtt::ConnectionInfo mqttInfo{url, user, pass};  ////   es::Mqtt::ConnectionInfo mqttInfo{url, mqttCert, user, pass};`

I have been trying dig deeper into the code, looking at the Espressif routines. I'm sure I could figure it out eventually, but decided to ask, in case I had missed something obvious.

**Comment by lubosmato** [comment]
Hi Rudy,

To use non-secured connection you can pass empty string_view as pem cert parameter and use `mqtt://` protocol in URL:

```cpp
es::Mqtt::ConnectionInfo mqttInfo{
  "mqtt://mybroker.com/",
  std::string_view{}, 
  user,
  pass
};
```

**Comment by RudyFiero** [comment]
I changed the line to 

`  es::Mqtt::ConnectionInfo mqttInfo{url, std::string_view{}, user, pass};  `

But I am still getting the following when  **wifi.connect(*ssid, *wifiPass);** is called.

```
[1B][0;32mI (695) wifi: disconnecting wifi[1B][0m
[1B][0;31mE (695) wifi_init: Failed to deinit Wi-Fi driver (0x3001)[1B][0m
[1B][0;32mI (695) wifi: connecting to wifi[1B][0m
I (715) wifi:wifi driver task: 3ffc1c5c, prio:23, stack:6656, core=0

***ERROR*** A stack overflow in task main has been detected.
```

I had commented out   wifi.connect(*ssid, *wifiPass);

That allowed the ESP32 to start in AP mode. I was able to put in the credentials.  
Then I re-enabled   wifi.connect(*ssid, *wifiPass); Did a reset and that still results in a stack overflow error.

Now I am wondering if this is because of the move to IDF 4.3? 
I am pretty sure that I ran the wifi.cpp example without error. I'll try it again to make sure.

**Comment by RudyFiero** [comment]
I started over. I setup a new project with the Espressif tcp mqtt example. It ran without error. connected to the network. I connected to the mqtt broker, and published messages. That example does not have  wifi.connect("My SSID", "my password");

I then made a new project with an expressif blink sketch. I modified it as per the instructions outlined in the documentation on the main page of this repository. I then copied the wifi.cpp to the main directory and renamed the file as main.cpp.  I then ran the code and still get the stack overflow errors. This code is as it appears here, no changes.  

I suspect that there is something messed up in the contents of the flash. Something that doesn't get changed when reprogramming the code. I have another ESP that I can try. Also I am going to try and do a complete wipe of the two I have been working with.

**Comment by RudyFiero** [comment]
I have been inching (2.54cm) my way through the code. I have been comparing the log output from mqtt.cpp and settings_server.cpp examples. My problems are in the mqtt section even though the stack overflows are happening when I have wifi.connect(*ssid, *wifiPass);  present. I have been putting a while(true){}; loop in place as a breakpoint. I am now thinking that I do not have the correct information entered for the mqtt broker. 

In the form I entered 192.168.1.68:1599  where now I think I need to add mqtt:// before the address ???

**Comment by RudyFiero** [comment]
I have got to the last segment in the example code. If I have a while loop in front of it, the program does not crash. I see mqtt messages posted to the broker. If I remove the while loop and allow the program to get to this code I get stack overflow errors. 

```
  int seconds = 0;
  while (true) {
    printf("myValue: %d\n", myValue);
    printf("myText: %s\n", myText.c_str());

    if (seconds % 10 == 0) {
      mqtt.publish("test/string", "how are you?", es::Mqtt::Qos::Qos0, false);
      mqtt.publish("test/integer", 42, es::Mqtt::Qos::Qos0, false);
      mqtt.publish("test/double", 42.4242, es::Mqtt::Qos::Qos0, false);
      mqtt.publish("test/bool", true, es::Mqtt::Qos::Qos0, false);
    }

    vTaskDelay(1000 / portTICK_PERIOD_MS);
    ++seconds;
  }
```

EDIT:  Commenting out one of the publish lines, then I get no error. First I had added delays between the sends, but that didn't help. Then I got rid of the delays but included only one publish. Then one more. Up to three worked. But the fourth was the straw that broke the camel's back.

```
  int seconds = 0;
  while (true) {
    printf("myValue: %d\n", myValue);
    printf("myText: %s\n", myText.c_str());
    if (seconds % 10 == 0) {
      mqtt.publish("test/string", "how are you?", es::Mqtt::Qos::Qos0, false);
      mqtt.publish("test/integer", 42, es::Mqtt::Qos::Qos0, false);
      mqtt.publish("test/double", 42.4242, es::Mqtt::Qos::Qos0, false);
//      mqtt.publish("test/bool", true, es::Mqtt::Qos::Qos0, false);
    }
    vTaskDelay(1000 / portTICK_PERIOD_MS);
    ++seconds;
  }
```

**Comment by RudyFiero** [comment]
Here is the output when it crashes, when there are four publish in a row. 

```
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3ffe6100,len:0x1828
load:0x4004c000,len:0x9e8
load:0x40050000,len:0x2b70
SHA-256 comparison failed:
Calculated: 7e0433fa95be6555ed6a6fa4b461bc6b9fb75f98a075669d6c29bd70526603d5
Expected: cfe9a17c90a4f87ae0f5167a586cf197a11fd4ec522f1fff32abf0f58c3b715a
Attempting to boot anyway...
entry 0x4004c1ec
[1B][0;32mI (40) boot: ESP-IDF v4.3-dirty 2nd stage bootloader[1B][0m
[1B][0;32mI (40) boot: compile time 10:55:07[1B][0m
[1B][0;32mI (40) boot: chip revision: 0[1B][0m
[1B][0;32mI (43) boot.esp32s2: SPI Speed      : 80MHz[1B][0m
[1B][0;32mI (47) boot.esp32s2: SPI Mode       : DIO[1B][0m
[1B][0;32mI (52) boot.esp32s2: SPI Flash Size : 4MB[1B][0m
[1B][0;32mI (57) boot: Enabling RNG early entropy source...[1B][0m
[1B][0;32mI (62) boot: Partition Table:[1B][0m
[1B][0;32mI (66) boot: ## Label            Usage          Type ST Offset   Length[1B][0m
[1B][0;32mI (73) boot:  0 nvs              WiFi data        01 02 00009000 00006000[1B][0m
[1B][0;32mI (81) boot:  1 phy_init         RF data          01 01 0000f000 00001000[1B][0m
[1B][0;32mI (88) boot:  2 factory          factory app      00 00 00010000 00100000[1B][0m
[1B][0;32mI (95) boot: End of partition table[1B][0m
[1B][0;32mI (100) esp_image: segment 0: paddr=00010020 vaddr=3f000020 size=26ff4h (159732) map[1B][0m
[1B][0;32mI (143) esp_image: segment 1: paddr=0003701c vaddr=3ffc6490 size=03e98h ( 16024) load[1B][0m
[1B][0;32mI (147) esp_image: segment 2: paddr=0003aebc vaddr=40022000 size=0515ch ( 20828) load[1B][0m
[1B][0;32mI (154) esp_image: segment 3: paddr=00040020 vaddr=40080020 size=9db08h (645896) map[1B][0m
[1B][0;32mI (296) esp_image: segment 4: paddr=000ddb30 vaddr=4002715c size=0f334h ( 62260) load[1B][0m
[1B][0;32mI (312) esp_image: segment 5: paddr=000ece6c vaddr=50000000 size=00010h (    16) load[1B][0m
[1B][0;32mI (324) boot: Loaded app from partition at offset 0x10000[1B][0m
[1B][0;32mI (324) boot: Disabling RNG early entropy source...[1B][0m
[1B][0;32mI (335) ca[03][06]žf†~[00]˜[18]†ž[1E]FÉÕÑ¥½¹cache 	: size 8KB, 4Ways, cache line size 32Byte[1B][0m
[1B][0;32mI (335) cpu_start: Pro cpu up.[1B][0m
[1B][0;32mI (349) cpu_start: Pro cpu start user code[1B][0m
[1B][0;32mI (349) cpu_start: cpu freq: 160000000[1B][0m
[1B][0;32mI (349) cpu_start: Application information:[1B][0m
[1B][0;32mI (354) cpu_start: Project name:     blink[1B][0m
[1B][0;32mI (359) cpu_start: App version:      1[1B][0m
[1B][0;32mI (363) cpu_start: Compile time:     Aug 12 2021 13:16:33[1B][0m
[1B][0;32mI (369) cpu_start: ELF file SHA256:  a915b3d2488f0a8b...[1B][0m
[1B][0;32mI (375) cpu_start: ESP-IDF:          v4.3-dirty[1B][0m
[1B][0;32mI (380) heap_init: Initializing. RAM available for dynamic allocation:[1B][0m
[1B][0;32mI (388) heap_init: At 3FF9E000 len 00002000 (8 KiB): RTCRAM[1B][0m
[1B][0;32mI (394) heap_init: At 3FFCEB50 len 0002D4B0 (181 KiB): DRAM[1B][0m
[1B][0;32mI (400) heap_init: At 3FFFC000 len 00003A10 (14 KiB): DRAM[1B][0m
[1B][0;32mI (407) spi_flash: detected chip: generic[1B][0m
[1B][0;32mI (411) spi_flash: flash io: dio[1B][0m
[1B][0;32mI (419) cpu_start: Starting scheduler on PRO CPU.[1B][0m
[1B][0;32mI (441) wifi: disconnecting wifi[1B][0m
[1B][0;31mE (441) wifi_init: Failed to deinit Wi-Fi driver (0x3001)[1B][0m
[1B][0;32mI (441) wifi: connecting to wifi[1B][0m
I (451) wifi:wifi driver task: 3ffd67bc, prio:23, stack:6656, core=0

***ERROR*** A stack overflow in task main has been detected.
```

And here is what I get when I comment out one of the publish.

```
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3ffe6100,len:0x1828
load:0x4004c000,len:0x9e8
load:0x40050000,len:0x2b70
SHA-256 comparison failed:
Calculated: 7e0433fa95be6555ed6a6fa4b461bc6b9fb75f98a075669d6c29bd70526603d5
Expected: cfe9a17c90a4f87ae0f5167a586cf197a11fd4ec522f1fff32abf0f58c3b715a
Attempting to boot anyway...
entry 0x4004c1ec
[1B][0;32mI (40) boot: ESP-IDF v4.3-dirty 2nd stage bootloader[1B][0m
[1B][0;32mI (40) boot: compile time 10:55:07[1B][0m
[1B][0;32mI (40) boot: chip revision: 0[1B][0m
[1B][0;32mI (43) boot.esp32s2: SPI Speed      : 80MHz[1B][0m
[1B][0;32mI (47) boot.esp32s2: SPI Mode       : DIO[1B][0m
[1B][0;32mI (52) boot.esp32s2: SPI Flash Size : 4MB[1B][0m
[1B][0;32mI (57) boot: Enabling RNG early entropy source...[1B][0m
[1B][0;32mI (62) boot: Partition Table:[1B][0m
[1B][0;32mI (66) boot: ## Label            Usage          Type ST Offset   Length[1B][0m
[1B][0;32mI (73) boot:  0 nvs              WiFi data        01 02 00009000 00006000[1B][0m
[1B][0;32mI (81) boot:  1 phy_init         RF data          01 01 0000f000 00001000[1B][0m
[1B][0;32mI (88) boot:  2 factory          factory app      00 00 00010000 00100000[1B][0m
[1B][0;32mI (95) boot: End of partition table[1B][0m
[1B][0;32mI (100) esp_image: segment 0: paddr=00010020 vaddr=3f000020 size=26fd8h (159704) map[1B][0m
[1B][0;32mI (143) esp_image: segment 1: paddr=00037000 vaddr=3ffc6490 size=03e98h ( 16024) load[1B][0m
[1B][0;32mI (147) esp_image: segment 2: paddr=0003aea0 vaddr=40022000 size=05178h ( 20856) load[1B][0m
[1B][0;32mI (154) esp_image: segment 3: paddr=00040020 vaddr=40080020 size=9daa0h (645792) map[1B][0m
[1B][0;32mI (296) esp_image: segment 4: paddr=000ddac8 vaddr=40027178 size=0f318h ( 62232) load[1B][0m
[1B][0;32mI (312) esp_image: segment 5: paddr=000ecde8 vaddr=50000000 size=00010h (    16) load[1B][0m
[1B][0;32mI (324) boot: Loaded app from partition at offset 0x10000[1B][0m
[1B][0;32mI (324) boot: Disabling RNG early entropy source...[1B][0m
[1B][0;32mI (335) ca[03][06]žf†~[00]˜[18]†ž[1E][06].®×Ñ¥½¹cache 	: size 8KB, 4Ways, cache line size 32Byte[1B][0m
[1B][0;32mI (335) cpu_start: Pro cpu up.[1B][0m
[1B][0;32mI (349) cpu_start: Pro cpu start user code[1B][0m
[1B][0;32mI (349) cpu_start: cpu freq: 160000000[1B][0m
[1B][0;32mI (349) cpu_start: Application information:[1B][0m
[1B][0;32mI (354) cpu_start: Project name:     blink[1B][0m
[1B][0;32mI (359) cpu_start: App version:      1[1B][0m
[1B][0;32mI (363) cpu_start: Compile time:     Aug 12 2021 13:16:33[1B][0m
[1B][0;32mI (369) cpu_start: ELF file SHA256:  fcf7d16432aee36b...[1B][0m
[1B][0;32mI (375) cpu_start: ESP-IDF:          v4.3-dirty[1B][0m
[1B][0;32mI (380) heap_init: Initializing. RAM available for dynamic allocation:[1B][0m
[1B][0;32mI (388) heap_init: At 3FF9E000 len 00002000 (8 KiB): RTCRAM[1B][0m
[1B][0;32mI (394) heap_init: At 3FFCEB50 len 0002D4B0 (181 KiB): DRAM[1B][0m
[1B][0;32mI (400) heap_init: At 3FFFC000 len 00003A10 (14 KiB): DRAM[1B][0m
[1B][0;32mI (407) spi_flash: detected chip: generic[1B][0m
[1B][0;32mI (411) spi_flash: flash io: dio[1B][0m
[1B][0;32mI (419) cpu_start: Starting scheduler on PRO CPU.[1B][0m
[1B][0;32mI (441) wifi: disconnecting wifi[1B][0m
[1B][0;31mE (441) wifi_init: Failed to deinit Wi-Fi driver (0x3001)[1B][0m
[1B][0;32mI (441) wifi: connecting to wifi[1B][0m
I (451) wifi:wifi driver task: 3ffd67bc, prio:23, stack:6656, core=0
[1B][0;32mI (451) system_api: Base MAC address is not set[1B][0m
[1B][0;32mI (451) system_api: read default base MAC address from EFUSE[1B][0m
I (461) wifi:wifi firmware version: c7d0450
I (461) wifi:wifi certification version: v7.0
I (461) wifi:config NVS flash: enabled
I (471) wifi:config nano formating: disabled
I (471) wifi:Init data frame dynamic rx buffer num: 32
I (481) wifi:Init management frame dynamic rx buffer num: 32
I (481) wifi:Init management short buffer num: 32
I (491) wifi:Init dynamic tx buffer num: 32
I (491) wifi:Init static rx buffer size: 1600
I (491) wifi:Init static rx buffer num: 10
I (501) wifi:Init dynamic rx buffer num: 32
[1B][0;32mI (501) wifi_init: rx ba win: 6[1B][0m
[1B][0;32mI (511) wifi_init: tcpip mbox: 32[1B][0m
[1B][0;32mI (511) wifi_init: udp mbox: 6[1B][0m
[1B][0;32mI (511) wifi_init: tcp mbox: 6[1B][0m
[1B][0;32mI (521) wifi_init: tcp tx win: 5744[1B][0m
[1B][0;32mI (521) wifi_init: tcp rx win: 5744[1B][0m
[1B][0;32mI (531) wifi_init: tcp mss: 1440[1B][0m
[1B][0;32mI (531) wifi_init: WiFi IRAM OP enabled[1B][0m
[1B][0;32mI (531) wifi_init: WiFi RX IRAM OP enabled[1B][0m
[1B][0;32mI (541) phy_init: phy_version 1800,e7ef680,Apr 13 2021,11:45:08[1B][0m
I (681) wifi:mode : sta (7c:df:a1:00:a6:1a)
I (681) wifi:enable tsf
[1B][0;32mI (681) wifi: trying to connect to AP SSID: 'Connect-2.4', password: 'AFX.ptc.ASC'...[1B][0m
Waiting for connection...
I (691) wifi:new:<7,0>, old:<1,0>, ap:<255,255>, sta:<7,0>, prof:5
I (691) wifi:state: init -> auth (b0)
I (701) wifi:state: auth -> assoc (0)
I (711) wifi:state: assoc -> run (10)
W (711) wifi:<ba-add>idx:0 (ifx:0, 60:32:b1:fd:6b:4f), tid:5, ssn:0, winSize:64
W (821) wifi:<ba-add>idx:1 (ifx:0, 60:32:b1:fd:6b:4f), tid:0, ssn:1, winSize:64
I (821) wifi:connected with Connect-2.4, aid = 7, channel 7, BW20, bssid = 60:32:b1:fd:6b:4f
I (821) wifi:security: WPA2-PSK, phy: bgn, rssi: -55
I (831) wifi:pm start, type: 1

I (851) wifi:AP's beacon interval = 102400 us, DTIM period = 1
[1B][0;32mI (1441) esp_netif_handlers: sta ip: 192.168.0.100, mask: 255.255.255.0, gw: 192.168.0.1[1B][0m
[1B][0;32mI (1441) wifi: connected[1B][0m
[1B][0;32mI (1441) wifi: got ip: 192.168.0.100[1B][0m
mqtt://192.168.1.78:1599 phasoninc phasonmosquitto
[1B][0;32mI (1691) mqtt: Free memory: 144692 bytes[1B][0m
myValue: 0
myText: 
MQTT is connected!
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
myValue: 0
myText: 
```

**Comment by lubosmato** [comment]
That looks very strange. In my projects I usually publish around 100 values every second and never had problems with stack overflow. I use esp32 wrover modules. 

Right now I have no option to test this. I will be able to test it after 23 of August. Could you upload/push somewhere your example code (with sdkconfig or ideally whole project) and tell me what esp32 module do you use?

**Comment by RudyFiero** [comment]
I had been using the ESP32-S2. This morning I made a new project and used an ESP32 WROOM module. Same thing is happening.

What confuses me is that the error appears to happen before it reaches that point in the code.  I am powering this off of the USB port. I am going to try it with the power coming from a power supply instead.

**Comment by RudyFiero** [comment]
I have an external power supply providing power to the ESP32-S2 board, and it still has the stack overflow error when I have the fourth publish enabled. I didn't think it would have been a power issue because of the repeatable nature of the failure. But I still needed to do it to eliminate power as a possible cause.

I had thought that maybe it was related to the processor being the ESP32-S2 part. Single core and less memory. But as I said in the above message, the problem still is there when I use a ESP32. 

I will still try and determine the reason for the failure. But as it is right now I can do additional work on the project. I need to include some I2C communications to a processor that I will be connecting to. 

I will get you a copy of the project with the ESP32. It might be here on github, or I might put it on Google Drive. I'll let you know.

**Comment by RudyFiero** [comment]
I changed the 1 second timer from 1000 to 20

 vTaskDelay(20 / portTICK_PERIOD_MS);

That publish task is now publishing 50 groups of 3 each second. So there is the throughput. But for some reason four in a row breaks it. I had thought putting delays between then would have helped, but it doesn't seem to matter. 

My tests are on the ESP32-S2 running at 240MHz.

**Comment by lubosmato** [comment]
Maybe the problem is exactly what error message says 🙂 - main task has not enough stack size. Could you try to set main task stack size (CONFIG_ESP_MAIN_TASK_STACK_SIZE) to bigger value? Eg. 10kB. 

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-esp-main-task-stack-size

**Comment by RudyFiero** [comment]
That was my thought as well. But I didn't know where I could change it. 
I will try an change it when I get back to this. If not this weekend, definitely on Monday.

**Comment by RudyFiero** [comment]
I changed it to 10000. I added two more publish commands in the same sequence. It is all working.

I did a Google on CONFIG_ESP_MAIN_TASK_STACK_SIZE and saw that the Arduino configuration changes this to something a bit higher than 10,000. 

I really appreciate that you have made this code available. When I had decided to use the native environment I thought that there would have been more examples to learn from. And thank you for your help.
